##Situational Overview (Business Need)
You are working for a wealthy investor that specialises in purchasing assets that are undervalued.  This investor’s due diligence on all purchases includes a detailed analysis of the data that underlies the business, to try to understand the fundamentals of the business and especially to identify opportunities to drive profitability by changing the focus of which products or services are being offered.

The investor is interested in purchasing TellCo, an existing mobile service provider in the Republic of Pefkakia.  TellCo’s current owners have been willing to share their financial information but have never employed anyone to look at their data that is generated automatically by their systems.

Your employer wants you to provide a report to analyse opportunities for growth and make a recommendation on whether TellCo is worth buying or selling.  You will do this by analysing a telecommunication dataset that contains useful information about the customers & their activities on the network. You will deliver insights you managed to extract to your employer through an easy to use web based dashboard and a written report. 

In [91]:
import numpy as np
import pandas as pd


In [92]:
summary=pd.read_excel('../data/Field Descriptions.xlsx')

In [93]:
#This table conntains information o the columns in our dataset and what they mean
summary.shape

(56, 2)

-There are 56 fields in our dataset.
-CDR or Call Detail Record is the voice channel. 
-XDR is the data channel. For example, one difference is that CDRs encode 2 phone numbers and two antennas (in and out) for a call, while XDRs only contain one antenna (the one you're connected to and downloading information)... no "out" antenna or "out" phone number is nedded. Another difference is perhaps that you're not measured in how many minutes you burn, but how many KBs you down/upload.

In [94]:
pd.set_option('display.max_columns', None)

In [95]:

summary

Fields  \
0                                  bearer id   
1                                  Dur. (ms)   
2                                      Start   
3                                   Start ms   
4                                        End   
5                                     End ms   
6                                   Dur. (s)   
7                                       IMSI   
8                              MSISDN/Number   
9                                       IMEI   
10                        Last Location Name   
11                           Avg RTT DL (ms)   
12                           Avg RTT UL (ms)   
13                   Avg Bearer TP DL (kbps)   
14                   Avg Bearer TP UL (kbps)   
15               TCP DL Retrans. Vol (Bytes)   
16               TCP UL Retrans. Vol (Bytes)   
17                       DL TP < 50 Kbps (%)   
18            50 Kbps < DL TP < 250 Kbps (%)   
19             250 Kbps < DL TP < 1 Mbps (%)   
20                        DL TP > 1 Mbps (%)   
21                       UL TP < 10 Kbps (%)   
22             10 Kbps < UL TP < 50 Kbps (%)   
23            50 Kbps < UL TP < 300 Kbps (%)   
24                      UL TP > 300 Kbps (%)   
25                           HTTP DL (Bytes)   
26                           HTTP UL (Bytes)   
27                 Activity Duration DL (ms)   
28                 Activity Duration UL (ms)   
29                               Dur. (ms).1   
30                      Handset Manufacturer   
31                              Handset Type   
32           Nb of sec with 125000B < Vol DL   
33     Nb of sec with 1250B < Vol UL < 6250B   
34  Nb of sec with 31250B < Vol DL < 125000B   
35            Nb of sec with 37500B < Vol UL   
36    Nb of sec with 6250B < Vol DL < 31250B   
37    Nb of sec with 6250B < Vol UL < 37500B   
38             Nb of sec with Vol DL < 6250B   
39             Nb of sec with Vol UL < 1250B   
40                   Social Media DL (Bytes)   
41                   Social Media UL (Bytes)   
42                        YouTube DL (Bytes)   
43                        YouTube UL (Bytes)   
44                        Netflix DL (Bytes)   
45                        Netflix UL (Bytes)   
46                         Google DL (Bytes)   
47                         Google UL (Bytes)   
48                          Email DL (Bytes)   
49                          Email UL (Bytes)   
50                         Gaming DL (Bytes)   
51                         Gaming UL (Bytes)   
52                                  Other DL   
53                                  Other UL   
54                          Total DL (Bytes)   
55                          Total UL (Bytes)   

                                          Description  
0                              xDr session identifier  
1                   Total Duration of the xDR (in ms)  
2       Start time of the xDR (first frame timestamp)  
3   Milliseconds offset of start time for the xDR ...  
4          End time of the xDR (last frame timestamp)  
5   Milliseconds offset of end time of the xDR (la...  
6                    Total Duration of the xDR (in s)  
7            International Mobile Subscriber Identity  
8   MS International PSTN/ISDN Number of mobile - ...  
9             International Mobile Equipment Identity  
10  User location call name (2G/3G/4G) at the end ...  
11  Average Round Trip Time measurement Downlink d...  
12  Average Round Trip Time measurement Uplink dir...  
13  Average Bearer Throughput for Downlink (kbps) ...  
14  Average Bearer Throughput for uplink (kbps) - ...  
15  TCP volume of Downlink packets detected as ret...  
16  TCP volume of Uplink packets detected as retra...  
17  Duration ratio when Bearer Downlink Throughput...  
18  Duration ratio when Bearer Downlink Throughput...  
19  Duration ratio when Bearer Downlink Throughput...  
20  Duration ratio when Bearer Downlink Throughput...  
21  Duration ratio when Bearer Uplink Throughput < ….  
22  Duration 

In [96]:
df=pd.read_csv('../data/Week1_challenge_data_source(CSV).csv',na_values=['?', None])


In [97]:

import pandas as pd
pd.set_option('display.float_format', lambda x: '%.3f' % x)

In [98]:
df.shape

(150001, 55)

There are 150,001 rows ad 55 columns

In [99]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
Bearer Id                                   149010 non-null float64
Start                                       150000 non-null object
Start ms                                    150000 non-null float64
End                                         150000 non-null object
End ms                                      150000 non-null float64
Dur. (ms)                                   150000 non-null float64
IMSI                                        149431 non-null float64
MSISDN/Number                               148935 non-null float64
IMEI                                        149429 non-null float64
Last Location Name                          148848 non-null object
Avg RTT DL (ms)                             122172 non-null float64
Avg RTT UL (ms)                             122189 non-null float64
Avg Bearer TP DL (kbps)                     150000 non-null float64
Avg Bear

In [100]:
df.describe(include=[np.number])

Bearer Id   Start ms     End ms   Dur. (ms)  \
count               149010.000 150000.000 150000.000  150000.000   
mean  10138874654256523264.000    499.188    498.801  104608.560   
std    2893172512268425216.000    288.612    288.098   81037.621   
min    6917537518543529984.000      0.000      0.000    7142.000   
25%    7349883246953990144.000    250.000    251.000   57440.500   
50%    7349883264156584960.000    499.000    500.000   86399.000   
75%   13042425978957574144.000    749.000    750.000  132430.250   
max   13186541167134199808.000    999.000    999.000 1859336.000   

                     IMSI       MSISDN/Number               IMEI  \
count          149431.000          148935.000         149429.000   
mean  208201639651672.219     41882819545.027 48474547977654.164   
std       21488090841.366   2447443358621.660 22416372027957.652   
min   204047108489451.000     33601001722.000   440015202000.000   
25%   208201401263249.000     33651295581.500 35460708865439.000   
50%   208201546329113.000     33663706799.000 35722009426311.000   
75%   208201771619103.000     33683490769.000 86119704674953.000   
max   214074303349628.000 882397108489451.000 99001201327774.000   

       Avg RTT DL (ms)  Avg RTT UL (ms)  Avg Bearer TP DL (kbps)  \
count       122172.000       122189.000               150000.000   
mean           109.796           17.663                13300.046   
std            619.783           84.794                23971.879   
min              0.000            0.000                    0.000   
25%             32.000            2.000                   43.000   
50%             45.000            5.000                   63.000   
75%             70.000           15.000                19710.750   
max          96923.000         7120.000               378160.000   

       Avg Bearer TP UL (kbps)  TCP DL Retrans. Vol (Bytes)  \
count               150000.000                    61855.000   
mean                  1770.429                 20809914.270   
std                   4625.356                182566527.915   
min                      0.000                        2.000   
25%                     47.000                    35651.500   
50%                     63.000                   568730.000   
75%                   1120.000                  3768308.500   
max                  58613.000               4294425570.000   

       TCP UL Retrans. Vol (Bytes)  DL TP < 50 Kbps (%)  \
count                    53352.000           149247.000   
mean                    759658.665               92.845   
std                   26453047.172               13.038   
min                          1.000                0.000   
25%                       4694.750               91.000   
50%                      20949.500              100.000   
75%                      84020.250              100.000   
max                 2908226006.000              100.000   

       50 Kbps < DL TP < 250 Kbps (%)  250 Kbps < DL TP < 1 Mbps (%)  \
count                      149247.000                     149247.000   
mean                            3.069                          1.717   
std                             6.215                          4.160   
min                             0.000                          0.000   
25%                             0.000                          0.000   
50%                             0.000                          0.000   
75%                             4.000                          1.000   
max                            93.000                        100.000   

       DL TP > 1 Mbps (%)  UL TP < 10 Kbps (%)  10 Kbps < UL TP < 50 Kbps (%)  \
count          149247.000           149209.000                     149209.000   
mean                1.610               98.530                          0.777   
std                 4.829                4.634                          3.225   
min                 0.000                0.000                          0.000   
25%                 0.000   

In [101]:
df.describe(include=[np.object])

Start             End Last Location Name  \
count           150000          150000             148848   
unique            9997            6403              45547   
top     4/26/2019 7:25  4/25/2019 0:01            D41377B   
freq               203            1150                 80   

       Handset Manufacturer      Handset Type  
count                149429            149429  
unique                  170              1396  
top                   Apple  Huawei B528S-23A  
freq                  59565             19752

# Renaming Column Names 

In [102]:
 #def renaming_columns( df:pd.DataFrame)->pd.DataFrame:
       
df.rename(columns = {'Bearer Id':'Bearer_Id'}, inplace = True)
df.rename(columns = {'Handset Manufacturer':'phone_company'}, inplace = True)
df.rename(columns = {'Handset Type':'phone_name'}, inplace = True)
df.rename(columns = {'MSISDN/Number':'MSISDN'}, inplace = True)
        
df.rename(columns = {'Last Location Name':'last_location'}, inplace = True)
              
df.rename(columns = {'Avg RTT UL (ms)':'avg_rtt_ul'}, inplace = True)
df.rename(columns = {'Avg RTT DL (ms)':'avg_rtt_dl'}, inplace = True)
        
df.rename(columns = {'Avg Bearer TP DL (kbps)':'throughput_avg_dl_kpbs'}, inplace = True)
df.rename(columns = {'Avg Bearer TP UL (kbps)':'throughput_avg_ul_kpbs'}, inplace = True)
        
df.rename(columns = {'TCP DL Retrans. Vol (Bytes)':'retrans_packets_dl_b'}, inplace = True)
df.rename(columns = {'TCP UL Retrans. Vol (Bytes)':'retrans_packets_ul_b'}, inplace = True)
        
df.rename(columns = {'DL TP < 50 Kbps (%)':'tp_dl_below_50kbps_pc'}, inplace = True)
df.rename(columns = {'50 Kbps < DL TP < 250 Kbps (%)':'tp_dl_50_250kbps_pc'}, inplace = True)
df.rename(columns = {'250 Kbps < DL TP < 1 Mbps (%)':'tp_dl_250kbps_1mbps_pc '}, inplace = True)
df.rename(columns = {'DL TP > 1 Mbps (%)':' tp_dl_above_1mbps_pc'}, inplace = True)        
        
df.rename(columns = {'UL TP < 10 Kbps (%)':'tp_ul_below_10kpbs_pc'}, inplace = True)
df.rename(columns = {'10 Kbps < UL TP < 50 Kbps (%)':'tp_ul_10_50_kbps_pc '}, inplace = True)
df.rename(columns = {'50 Kbps < UL TP < 300 Kbps (%)':'tp_ul_50_300_kbps_pc  '}, inplace = True)
df.rename(columns = {'UL TP > 300 Kbps (%)':'  tp_ul_above_300_kpbs_pc'}, inplace = True)


        
df.rename(columns = {'HTTP DL (Bytes)':'http_dl_b'}, inplace = True)
df.rename(columns = {'HTTP UL (Bytes)':'http_ul_b'}, inplace = True)
        
df.rename(columns = {'Activity Duration DL (ms)':'activity_duration_dl'}, inplace = True)
df.rename(columns = {'Activity Duration UL (ms)':'activity_duration_ul'}, inplace = True)
df.rename(columns = {'Activity Duration UL (ms)':'activity_duration_ul'}, inplace = True)
        
df.rename(columns = {'Nb of sec with 125000B < Vol DL':'t_vol_dl_above_125000B '}, inplace = True)
df.rename(columns = {'Nb of sec with 1250B < Vol UL < 6250B':'t_vol_ul_1250B_6250B  '}, inplace = True)
        
df.rename(columns = {'Nb of sec with 31250B < Vol DL < 125000B':'t_vol_dl_31250B_125000B'}, inplace = True)
df.rename(columns = {'Nb of sec with 37500B < Vol UL':'t_vol_ul_above_37500B'}, inplace = True)
        
df.rename(columns = {'Nb of sec with 6250B < Vol DL < 31250B':' t_vol_dl_6250B_31250B'}, inplace = True)
df.rename(columns = {'Nb of sec with 6250B < Vol UL < 37500B':'t_vol_ul_6250_37500B'}, inplace = True)
        
df.rename(columns = {'Nb of sec with Vol DL < 6250B':'t_vol_dl_above_6250B '}, inplace = True)
df.rename(columns = {'Nb of sec with Vol UL < 1250B':'t_vol_ul_above_1250B'}, inplace = True)
                                  
        
df.rename(columns = {'Social Media UL (Bytes)':'socials_ul_b'}, inplace = True)
df.rename(columns = {'Social Media DL (Bytes)':'socials_dl_b'}, inplace = True)
        
df.rename(columns = {'Google UL (Bytes)':'google_ul_b'}, inplace = True)
df.rename(columns = {'Google DL (Bytes)':'google_dl_b'}, inplace = True)
        
df.rename(columns = {'Email UL (Bytes)':'email_ul_b'}, inplace = True)
df.rename(columns = {'Email DL (Bytes)':'email_dl_b'}, inplace = True)
        
df.rename(columns = {'Youtube UL (Bytes)':'youtube_ul_b'}, inplace = True)
df.rename(columns = {'Youtube DL (Bytes)':'youtube_dl_b'}, inplace = True)
        
df.rename(columns = {'Netflix UL (Bytes)':'netflix_ul_b'}, inplace = True)
df.rename(columns = {'Netflix DL (Bytes)':'netflix_dl_b'}, inplace = True)
        
df.rename(columns = {'Gaming UL (Bytes)':'gaming_ul_b'}, inplace = True)
df.rename(columns = {'Gaming DL (Bytes)':'gaming_dl_b'}, inplace = True)
        
df.rename(columns = {'Other UL (Bytes)':'other_ul_b'}, inplace = True)
df.rename(columns = {'Other DL (Bytes)':'other_dl_b'}, inplace = True)
        
        
df.rename(columns = {'Total UL (Bytes)':'Total_ul_b'}, inplace = True)
df.rename(columns = {'Total DL (Bytes)':'Total_dl_b'}, inplace = True)
        
        

        

In [103]:
df.columns

Index(['Bearer_Id', 'Start', 'Start ms', 'End', 'End ms', 'Dur. (ms)', 'IMSI',
       'MSISDN', 'IMEI', 'last_location', 'avg_rtt_dl', 'avg_rtt_ul',
       'throughput_avg_dl_kpbs', 'throughput_avg_ul_kpbs',
       'retrans_packets_dl_b', 'retrans_packets_ul_b', 'tp_dl_below_50kbps_pc',
       'tp_dl_50_250kbps_pc', 'tp_dl_250kbps_1mbps_pc ',
       ' tp_dl_above_1mbps_pc', 'tp_ul_below_10kpbs_pc',
       'tp_ul_10_50_kbps_pc ', 'tp_ul_50_300_kbps_pc  ',
       '  tp_ul_above_300_kpbs_pc', 'http_dl_b', 'http_ul_b',
       'activity_duration_dl', 'activity_duration_ul', 'Dur. (ms).1',
       'phone_company', 'phone_name', 't_vol_dl_above_125000B ',
       't_vol_ul_1250B_6250B  ', 't_vol_dl_31250B_125000B',
       't_vol_ul_above_37500B', ' t_vol_dl_6250B_31250B',
       't_vol_ul_6250_37500B', 't_vol_dl_above_6250B ', 't_vol_ul_above_1250B',
       'socials_dl_b', 'socials_ul_b', 'google_dl_b', 'google_ul_b',
       'email_dl_b', 'email_ul_b', 'youtube_dl_b', 'youtube_ul_b',
       'ne

In [104]:
#To save on memory resources we convert the object data types into category

In [105]:

        
df["phone_company"] = df["phone_company"].astype("category")
df["phone_name"] = df["phone_name"].astype("category")
df["last_location"] = df["last_location"].astype("category")
df["Bearer_Id"] = df["Bearer_Id"].astype("category")
df["IMSI"] = df["IMSI"].astype("category")
df["MSISDN"] = df["MSISDN"].astype("category")
df["IMEI"] = df["IMEI"].astype("category")
       
        
        
      

-There are 13,4709 unique bearers in the dataset
-170 handset manufacturers with Apple being the most common
-What more information can we get from the location name information?

In [106]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
Bearer_Id                    149010 non-null category
Start                        150000 non-null object
Start ms                     150000 non-null float64
End                          150000 non-null object
End ms                       150000 non-null float64
Dur. (ms)                    150000 non-null float64
IMSI                         149431 non-null category
MSISDN                       148935 non-null category
IMEI                         149429 non-null category
last_location                148848 non-null category
avg_rtt_dl                   122172 non-null float64
avg_rtt_ul                   122189 non-null float64
throughput_avg_dl_kpbs       150000 non-null float64
throughput_avg_ul_kpbs       150000 non-null float64
retrans_packets_dl_b         61855 non-null float64
retrans_packets_ul_b         53352 non-null float64
tp_dl_below_50kbps_pc        149247 non-

# Dropping columns

In [107]:
# column_names = df.columns.to_list()
num_missing = df.isnull().sum()
shape = df.shape
num_rows = df.shape[0]
# num_cells = np.product(df.shape)

data = {
    # 'column_names': column_names, 
    'num_missing': num_missing, 
    'percent_missing (%)': [round(x,2) for x in num_missing/num_rows*100]
}

stats = pd.DataFrame(data)

In [108]:
stats[stats['num_missing'] != 0]

num_missing  percent_missing (%)
Bearer_Id                          991                0.660
Start                                1                0.000
Start ms                             1                0.000
End                                  1                0.000
End ms                               1                0.000
Dur. (ms)                            1                0.000
IMSI                               570                0.380
MSISDN                            1066                0.710
IMEI                               572                0.380
last_location                     1153                0.770
avg_rtt_dl                       27829               18.550
avg_rtt_ul                       27812               18.540
throughput_avg_dl_kpbs               1                0.000
throughput_avg_ul_kpbs               1                0.000
retrans_packets_dl_b             88146               58.760
retrans_packets_ul_b             96649               64.430
tp_dl_below_50kbps_pc              754                0.500
tp_dl_50_250kbps_pc                754                0.500
tp_dl_250kbps_1mbps_pc             754                0.500
 tp_dl_above_1mbps_pc              754                0.500
tp_ul_below_10kpbs_pc              792                0.530
tp_ul_10_50_kbps_pc                792                0.530
tp_ul_50_300_kbps_pc               792                0.530
  tp_ul_above_300_kpbs_pc          792                0.530
http_dl_b                        81474               54.320
http_ul_b                        81810               54.540
activity_duration_dl                 1                0.000
activity_duration_ul                 1                0.000
Dur. (ms).1                          1                0.000
phone_company                      572                0.380
phone_name                         572                0.380
t_vol_dl_above_125000B           97538               65.020
t_vol_ul_1250B_6250B             92894               61.930
t_vol_dl_31250B_125000B          93586               62.390
t_vol_ul_above_37500B           130254               86.840
 t_vol_dl_6250B_31250B           88317               58.880
t_vol_ul_6250_37500B            111843               74.560
t_vol_dl_above_6250B               755                0.500
t_vol_ul_above_1250B               793                0.530
Total_ul_b                           1                0.000
Total_dl_b                           1                0.000

In [109]:
#large_missing = ['weight', 'payer_code', 'medical_specialty', 'max_glu_serum', 'A1Cresult']
large_missing = stats[stats['percent_missing (%)'] > 65].index.to_list()
print(large_missing)

['t_vol_dl_above_125000B ', 't_vol_ul_above_37500B', 't_vol_ul_6250_37500B']


In [110]:
#let us drop the columns we will ot need
df = df.drop(large_missing, axis=1)

In [111]:


df=df.drop(columns=['Start', 'Start ms', 'End', 'End ms','http_dl_b', 'http_ul_b', 't_vol_ul_1250B_6250B  ', 't_vol_dl_31250B_125000B',  ' t_vol_dl_6250B_31250B',])


# #missing values

In [112]:
# get the number of missing data points per column
missing_values_count = df.isnull().sum()

# look at the # of missing points in the first ten columns
missing_values_count

Bearer_Id                      991
Dur. (ms)                        1
IMSI                           570
MSISDN                        1066
IMEI                           572
last_location                 1153
avg_rtt_dl                   27829
avg_rtt_ul                   27812
throughput_avg_dl_kpbs           1
throughput_avg_ul_kpbs           1
retrans_packets_dl_b         88146
retrans_packets_ul_b         96649
tp_dl_below_50kbps_pc          754
tp_dl_50_250kbps_pc            754
tp_dl_250kbps_1mbps_pc         754
 tp_dl_above_1mbps_pc          754
tp_ul_below_10kpbs_pc          792
tp_ul_10_50_kbps_pc            792
tp_ul_50_300_kbps_pc           792
  tp_ul_above_300_kpbs_pc      792
activity_duration_dl             1
activity_duration_ul             1
Dur. (ms).1                      1
phone_company                  572
phone_name                     572
t_vol_dl_above_6250B           755
t_vol_ul_above_1250B           793
socials_dl_b                     0
socials_ul_b        

In [113]:
# how many total missing values do we have?
total_cells = np.product(df.shape)
total_missing = missing_values_count.sum()

# percent of data that is missing
(total_missing/total_cells) * 100

3.93287300565283

In [114]:
#Letus drop the last row with missing values

In [115]:
df = df.dropna(axis=0, subset=['Total_dl_b'])

In [116]:
df = df.dropna(axis=0, subset=['MSISDN'])

In [121]:
df = df.dropna(axis=0, subset=['Bearer_Id'])

In [122]:
df.shape

(148506, 43)

In [123]:
# get the number of missing data points per column
missing_values_count = df.isnull().sum()

# look at the # of missing points in the first ten columns
missing_values_count

Bearer_Id                      0
Dur. (ms)                      0
IMSI                           0
MSISDN                         0
IMEI                           0
last_location                160
avg_rtt_dl                     0
avg_rtt_ul                     0
throughput_avg_dl_kpbs         0
throughput_avg_ul_kpbs         0
retrans_packets_dl_b           0
retrans_packets_ul_b           0
tp_dl_below_50kbps_pc        712
tp_dl_50_250kbps_pc          712
tp_dl_250kbps_1mbps_pc       712
 tp_dl_above_1mbps_pc        712
tp_ul_below_10kpbs_pc        767
tp_ul_10_50_kbps_pc          767
tp_ul_50_300_kbps_pc         767
  tp_ul_above_300_kpbs_pc    767
activity_duration_dl           0
activity_duration_ul           0
Dur. (ms).1                    0
phone_company                  0
phone_name                     0
t_vol_dl_above_6250B         713
t_vol_ul_above_1250B         768
socials_dl_b                   0
socials_ul_b                   0
google_dl_b                    0
google_ul_

In [118]:
df.describe().T

count          mean           std         min  \
Dur. (ms)                 148935.000    104870.163     81063.429    7142.000   
avg_rtt_dl                121291.000       108.225       594.300       0.000   
avg_rtt_ul                121310.000        17.637        84.793       0.000   
throughput_avg_dl_kpbs    148935.000     13286.171     23978.591       0.000   
throughput_avg_ul_kpbs    148935.000      1770.786      4629.476       0.000   
retrans_packets_dl_b       61139.000  20884183.275 182594305.983       2.000   
retrans_packets_ul_b       52743.000    766247.496  26604599.228       1.000   
tp_dl_below_50kbps_pc     148215.000        92.864        13.028       0.000   
tp_dl_50_250kbps_pc       148215.000         3.058         6.207       0.000   
tp_dl_250kbps_1mbps_pc    148215.000         1.714         4.158       0.000   
 tp_dl_above_1mbps_pc     148215.000         1.608         4.822       0.000   
tp_ul_below_10kpbs_pc     148168.000        98.533         4.638       0.000   
tp_ul_10_50_kbps_pc       148168.000         0.775         3.224       0.000   
tp_ul_50_300_kbps_pc      148168.000         0.148         1.627       0.000   
  tp_ul_above_300_kpbs_pc 148168.000         0.079         1.299       0.000   
activity_duration_dl      148935.000   1834541.749   5710292.649       0.000   
activity_duration_ul      148935.000   1412540.125   4655541.146       0.000   
Dur. (ms).1               148935.000 104870707.986  81063418.631 7142988.000   
t_vol_dl_above_6250B      148214.000      3728.048      9190.893       1.000   
t_vol_ul_above_1250B      148167.000      4029.250     10179.921       1.000   
socials_dl_b              148935.000   1795161.287   1035457.644      12.000   
socials_ul_b              148935.000     32921.716     19006.779       0.000   
google_dl_b               148935.000   5751574.185   3309040.749     207.000   
google_ul_b               148935.000   2056217.216   1189936.311       3.000   
email_dl_b                148935.000   1791768.717   1035909.953      14.000   
email_ul_b                148935.000    467389.474    269974.713       2.000   
youtube_dl_b              148935.000  11633449.989   6711383.509      53.000   
youtube_ul_b              148935.000  11008671.160   6345427.973     105.000   
netflix_dl_b              148935.000  11627558.354   6724685.556      42.000   
netflix_ul_b              148935.000  11000165.399   6358918.107      35.000   
gaming_dl_b               148935.000 422025598.218 243976284.233    2516.000   
gaming_ul_b               148935.000   8289085.431   4782090.168      59.000   
other_dl_b                148935.000 421147308.853 243197032.398    3290.000   
other_ul_b                148935.000   8263189.121   4768708.127     148.000   
Total_ul_b                148935.000  41117639.517  11279332.064 2866892.000   
Total_dl_b                148935.000 454625110.750 244152774.242 7114041.000   

                                    25%           50%           75%  \
Dur. (ms)                     57883.500     86399.000    132700.000   
avg_rtt_dl                       32.000        45.000        69.000   
avg_rtt_ul                        2.000         5.000        15.000   
throughput_avg_dl_kpbs           43.000        63.000     19681.000   
throughput_avg_ul_kpbs           47.000        63.000      1117.000   
retrans_packets_dl_b          36098.500    576130.000   3790785.000   
retrans_packets_ul_b           4721.500     21121.000     84695.500   
tp_dl_below_50kbps_pc            91.000       100.000       100.000   
tp_dl_50_250kbps_pc               0.000         0.000         4.000   
tp_dl_250kbps_1mbps_pc            0.000         0.000         1.000   
 tp_dl_above_1mbps_pc             0.000         0.000         0.000   
tp_ul_below_10kpbs_pc            99.000       100.000       100.000   
tp_ul_10_50_kbps_pc               0.000         0.000         0.000   
tp_ul_50_300_kbps_pc              0.000         0.000         0.000   
  tp_

In [124]:
df.loc[np.isnan(df["avg_rtt_dl"]), 'avg_rtt_dl'] = 108.225
df.loc[np.isnan(df["avg_rtt_ul"]), 'avg_rtt_ul'] = 17.637
df.loc[np.isnan(df["retrans_packets_dl_b"]), 'retrans_packets_dl_b'] = 20884183.275
df.loc[np.isnan(df["retrans_packets_ul_b"]), 'retrans_packets_ul_b'] = 766247.496

In [126]:
#df.loc[np.isnan(df["t_vol_dl_above_6250B"]), 't_vol_dl_above_6250B'] = 202
df.loc[np.isnan(df["t_vol_ul_above_1250B"]), 't_vol_ul_above_1250B'] = 4029.250
df.loc[np.isnan(df["tp_ul_below_10kpbs_pc"]), 'tp_ul_below_10kpbs_pc'] = 98.533         


In [131]:
# get the number of missing data points per column
missing_values_count = df.isnull().sum()

# look at the # of missing points in the first ten columns
missing_values_count

Bearer_Id                      0
Dur. (ms)                      0
IMSI                           0
MSISDN                         0
IMEI                           0
last_location                160
avg_rtt_dl                     0
avg_rtt_ul                     0
throughput_avg_dl_kpbs         0
throughput_avg_ul_kpbs         0
retrans_packets_dl_b           0
retrans_packets_ul_b           0
tp_dl_below_50kbps_pc        712
tp_dl_50_250kbps_pc          712
tp_dl_250kbps_1mbps_pc       712
 tp_dl_above_1mbps_pc        712
tp_ul_below_10kpbs_pc          0
tp_ul_10_50_kbps_pc          767
tp_ul_50_300_kbps_pc         767
  tp_ul_above_300_kpbs_pc    767
activity_duration_dl           0
activity_duration_ul           0
Dur. (ms).1                    0
phone_company                  0
phone_name                     0
t_vol_dl_above_6250B         713
t_vol_ul_above_1250B           0
socials_dl_b                   0
socials_ul_b                   0
google_dl_b                    0
google_ul_

Let us explore the dataset to see how to cater to missing values

-IMEI is a 15 digit number assigned to every cellular device for each of its SIM slots;
-IMSI is a 15 digit number assigned to the SIM of a mobile subscriber;
-MSISDN is the full mobile number including the country code and any prefixes.

Dropping all nans is not a good idea.?We therefore have to perform imputation for missing value treatment

In [132]:
data=df.fillna(method="ffill")

In [133]:
# get the number of missing data points per column
missing_values_count = data.isnull().sum()

# look at the # of missing points in the first ten columns
missing_values_count

Bearer_Id                    0
Dur. (ms)                    0
IMSI                         0
MSISDN                       0
IMEI                         0
last_location                0
avg_rtt_dl                   0
avg_rtt_ul                   0
throughput_avg_dl_kpbs       0
throughput_avg_ul_kpbs       0
retrans_packets_dl_b         0
retrans_packets_ul_b         0
tp_dl_below_50kbps_pc        0
tp_dl_50_250kbps_pc          0
tp_dl_250kbps_1mbps_pc       0
 tp_dl_above_1mbps_pc        0
tp_ul_below_10kpbs_pc        0
tp_ul_10_50_kbps_pc          0
tp_ul_50_300_kbps_pc         0
  tp_ul_above_300_kpbs_pc    0
activity_duration_dl         0
activity_duration_ul         0
Dur. (ms).1                  0
phone_company                0
phone_name                   0
t_vol_dl_above_6250B         0
t_vol_ul_above_1250B         0
socials_dl_b                 0
socials_ul_b                 0
google_dl_b                  0
google_ul_b                  0
email_dl_b                   0
email_ul

-ffill() is equivalent to fillna(method='ffill') and bfill() is equivalent to fillna(method='bfill')
-pad / ffill -Fill values forward.This method would fill the missing values with first non-missing value that occurs before it:
-bfill / backfill -Fill values backward.This method would fill the missing values with first non-missing value that occurs after it:

In [134]:
# remove all columns with at least one missing value
cleaned_data = data.dropna(axis=0)
cleaned_data.shape

(148506, 43)

In [135]:
cleaned_data.reset_index(inplace = True)

We drop around 11 rows.Let us convert the dataset to a csv file

In [136]:
cleaned_data.to_csv('../data/clean_data.csv', encoding='utf-8')